pkg/plugins/plugins.go

```diff
@@ -252,21 +266,53 @@ func (scanner *PluginScanner) loadPluginJson(pluginJsonFilePath string) error {
 	}
 
 	if pluginCommon.Id == "" || pluginCommon.Type == "" {
-		return errors.New("Did not find type and id property in plugin.json")
+		return errors.New("did not find type or id properties in plugin.json")
+	}
+
+	pluginCommon.PluginDir = filepath.Dir(pluginJsonFilePath)
+
+	if scanner.requireSigned {
+		scanner.log.Debug("Plugin signature required, validating", "pluginID", pluginCommon.Id,
+			"pluginDir", pluginCommon.PluginDir)
+		allowUnsigned := false
+		for _, plug := range scanner.cfg.PluginsAllowUnsigned {
+			if plug == pluginCommon.Id {
+				allowUnsigned = true
+				break
+			}
+		}
+		if sig := GetPluginSignatureState(&pluginCommon); sig != PluginSignatureValid && !allowUnsigned {
+			switch sig {
+			case PluginSignatureUnsigned:
+				scanner.log.Debug("Cannot load plugin since it has no signature", "pluginID", pluginCommon.Id)
+				return fmt.Errorf("plugin %q is unsigned", pluginCommon.Id)
+			case PluginSignatureInvalid:
+				scanner.log.Debug("Cannot load plugin since it has an invalid signature", "pluginID", pluginCommon.Id)
+				return fmt.Errorf("plugin %q has an invalid signature", pluginCommon.Id)
+			case PluginSignatureModified:
+				scanner.log.Debug("Cannot load plugin since its signature is modified", "pluginID", pluginCommon.Id)
+				return fmt.Errorf("plugin %q's signature has been modified", pluginCommon.Id)
+			default:
+				panic(fmt.Sprintf("Unrecognized plugin signature state %v", sig))
```

Comment:
do we really want to panic?

Reply:
@ryantxu I _think_ a panic makes sense here, since it means that at some point another possible plugin signature state has been added to the code, and we don't know what to do with it, i.e. it's a programmer error. In C/++ I would use an assert for this, whereas in Rust the compiler catches it :)

Reply:
I think panic will crash the whole Grafana process. Pretty sure we don't want to do that. Why not return an error not load the plugin?

Reply:
since this is user input. Panic is a bit harsh.

Reply:
@bergquist @kylebrandt @ryantxu OK, I'll bow to the majority and turn it into an error.